### PR TITLE
Allow SALT in CREATE USER user1 IDENTIFIED WITH sha256_hash

### LINKED ISSUE #37331: CREATE USER fails with "ON CLUSTER" and "sha256_password"
I'm trying to create a new user on a clickhouse cluster with  3 shards using this command:
`CREATE USER IF NOT EXISTS user ON CLUSTER '{cluster}' IDENTIFIED WITH sha256_password BY 'password' DEFAULT ROLE NONE`

This command gets expanded to:

`CREATE USER IF NOT EXISTS user ON CLUSTER {cluster} IDENTIFIED WITH sha256_hash BY '<hash>' SALT '<salt>' DEFAULT ROLE NONE`

However, on each of the shard I get this error:

`Syntax error: failed at position 123 ('SALT'): SALT '<salt>' DEFAULT ROLE NONE. Expected one of: HOST, SETTINGS, DEFAULT ROLE, GRANTEES, DEFAULT DATABASE, end of query.`

However, if I switch to use `plaintext_password` instead, everything works fine. Looks like the expansion and the addition of the "SALT" was not expected when handling this command.

* Which ClickHouse server version to use: 22.4

**Expected behavior**
The user to be created with no errors.

Thanks!

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
SALT is allowed for CREATE USER <user> IDENTIFIED WITH sha256_hash

closes #37331

### CODE REVIEW

File: src/Parsers/Access/ParserCreateUserQuery.cpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ namespace
     }
 
 
-    bool parseAuthenticationData(IParserBase::Pos & pos, Expected & expected, bool id_mode, AuthenticationData & auth_data)
+    bool parseAuthenticationData(IParserBase::Pos & pos, Expected & expected, bool /*id_mode*/, AuthenticationData & auth_data)
```
Comment (tavplubix) [inline src/Parsers/Access/ParserCreateUserQuery.cpp:49]: Argument `id_mode` is unused now, consider removing it

**Review by tavplubix — COMMENTED** [review]
Please add a test

**Review by tavplubix — APPROVED** [review]

File: src/Parsers/Access/ParserCreateUserQuery.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ namespace
                     return false;
                 value = ast->as<const ASTLiteral &>().value.safeGet<String>();
 
-                if (id_mode && expect_hash)
+                if (expect_hash)
```
Comment (vitlibar) [inline src/Parsers/Access/ParserCreateUserQuery.cpp:123]: This condition is a bit inconsistent. Salt can be used only with `SHA256_PASSWORD`, but here we try to parse it for `DOUBLE_SHA1_HASH` too.

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: Actually expect_hash will be set for SHA256_HASH, but let me check DOUBLE_SHA1_HASH...

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: I don't see `expect_hash` will be set for `DOUBLE_SHA1_HASH` - in corresponding code `type` will be set in previous loop and code with setting `expect_hash` seems redundant to me...

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: correction
now `SALT` can be used for both - SHA256_HASH and DOUBLE_SHA1_HASH with the same effect.

Comment (vitlibar) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: I meant salt is not supported for DOUBLE_SHA1_HASH (and probably doesn't need to be supported) and better not to parse it here

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: I would look at it other way around - currently salt is not added for `DOUBLE_SHA1_PASSWORD` - but maybe it should? and then salt for `DOUBLE_SHA1_HASH` will be more consistent. But anyway I see possibility to set salt for `DOUBLE_SHA1_HASH` as a feature.

Comment (vitlibar) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: DOUBLE_SHA1 is often used with MySQL native protocol implementation which can't work with salt.
And if something is not going to work for me it's better to tell user about it.

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: what the scenario of using of our DOUBLE_SHA1 with MySQL? Let's say we made `CREATE USER ... IDENTIFIED WITH double _sha1_**password**` - how this is used with MySQL?

Comment (vitlibar) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: ClickHouse server can pretend to be MySQL server so if we enable that in the configuration any MySQL client can connect to ClickHouse server thinking it's a MySQL server via MySQL native protocol. For that connection one pass a user name and a password and they will be checked by ClickHouse if the corresponding user is defined `WITH double_sha1_password`

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: Then I think there is no problem - salt is completely internal thing and transparent when password is used - salt is added to provided password internally, then this concatenation is hashed and compared with saved hash.
I can provide more information on why and how salt is used if you wish to make sure we are on the same page and I don't make any wrong conclusions...

Comment (vitlibar) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: MySQL clients don't send plaintext password, they send hash instead. To apply salt we need plaintext password, we cannot apply salt to what already hashed.

Comment (yakov-olkhovskiy) [reply src/Parsers/Access/ParserCreateUserQuery.cpp:123]: ...of course I can remove it for consistency... what do you think will be the best?

**Comment by yakov-olkhovskiy** [comment]
I decided to make it consistent, so no SALT allowed for DOUBLE_SHA1_HASH

**Pull request merged**